src/django/api/views/boundary.py

```diff
@@ -0,0 +1,84 @@
+from django.db.models import Prefetch
+from django.shortcuts import get_object_or_404
+
+from rest_framework.response import Response
+from rest_framework.exceptions import APIException
+from rest_framework.views import APIView
+from rest_framework.permissions import IsAuthenticated
+
+from ..models.user import Roles
+from ..models.boundary import Boundary
+from ..models.utility import Utility
+from ..models.submission import Submission
+
+from ..serializers.boundary import BoundaryDetailSerializer, BoundaryListSerializer
+
+
+def get_boundary_queryset_for_user(user):
+    if user.role == Roles.CONTRIBUTOR:
+        return Boundary.objects.filter(utility__in=user.utilities.all())
+
+    if user.role == Roles.VALIDATOR:
+        return Boundary.objects.filter(utility__state__in=user.states.all())
+
+    if user.role == Roles.ADMINISTRATOR:
+        return Boundary.objects.all()
+
+    raise RuntimeError('Invalid role: {}'.format(user.role))
+
+
+class BoundaryListView(APIView):
+    permission_classes = [IsAuthenticated]
+
+    def get(self, request, format=None):
+        boundaries = get_boundary_queryset_for_user(request.user)
+        boundaries = boundaries.select_related('utility')
+
+        requested_utilities = BoundaryListView.get_requested_utilities(request)
+        if requested_utilities:
+            boundaries = boundaries.filter(utility__in=requested_utilities)
+
+        return Response(BoundaryListSerializer(boundaries, many=True).data)
+
+    @staticmethod
+    def get_requested_utilities(request):
+        utility_ids = BoundaryListView.get_requested_utility_ids(request)
+        if not utility_ids:
+            return None
+
+        utilities = []
+        for utility_id in utility_ids:
+            utility = get_object_or_404(Utility, pk=utility_id)
+
+            if not request.user.has_access_to_utility(utility):
+                raise APIException('Cannot access this utility', 401)
+
+            utilities.append(utility)
```

Comment:
Consider something like:

```python
if user.role == ADMINISTRATOR:
  utilities = Utilities.objects.filter(id__in=utility_ids)

if user.role == VALIDATOR:
  utilities = user.states.utilities.filter(id__in=utility_ids)

if user.role == CONTRIBUTOR:
  utilities = user.utilities.filter(id__in=utility_ids)
```

Which would get multiple utilities from the database in one go, instead of getting them one at a time with `get_object_or_404`.

Reply:
Also, we should return an empty list for non-matching utilities, instead of 404s and 401s.

Reply:
These checks should be redundant because the original query set filters on user permissions. Not requiring an erro means we can just use the IDs from the query parameters directly.

8ce625e6efb687e7e3b135821721a71219125e5a